device/core/src/device_method/method_client.ts

```diff
@@ -0,0 +1,115 @@
+        this._restApiClient.executeApiCall('POST', path, this._httpHeaders, body, methodTimeout, (err, result) => {
+          if (err) {
+            /*Codes_SRS_NODE_DEVICE_METHOD_CLIENT_16_013: [The `invokeMethod` method shall call its callback with an error if `RestApiClient.executeApiCall` fails.]*/
+            callback(err);
+          } else {
+            /*Codes_SRS_NODE_DEVICE_METHOD_CLIENT_16_014: [The `invokeMethod` method shall call its callback with the result object if the call to `RestApiClient.executeApiCall` succeeds.]*/
+            callback(null, result as MethodResult);
+          }
+        });
+      }
+    });
+  }
+
+  setOptions(options: any): void {
+    /*Codes_SRS_NODE_DEVICE_METHOD_CLIENT_16_001: [The `setOptions` method shall merge the options passed in argument with the existing set of options used by the `MethodClient`.]*/
+    this._options = _.merge(this._options, options);
+  }
+
+  updateSharedAccessSignature(sharedAccessSignature: string, callback: (err?: Error) => void): void {
+    /*Codes_SRS_NODE_DEVICE_METHOD_CLIENT_16_002: [The `updateSharedAccessSignature` method shall create the `RestApiClient` object used by the client if it hasn't been created yet.]*/
+    this._init((err) => {
+      if (err) {
+        /*Codes_SRS_NODE_DEVICE_METHOD_CLIENT_16_003: [The `updateSharedAccessSignature` method shall call its callback with an Error if it fails to get the credentials from the `AuthenticationProvider` in order to create the `RestApiClient`.]*/
+        callback(err);
+      } else {
+        /*Codes_SRS_NODE_DEVICE_METHOD_CLIENT_16_005: [The `updateSharedAccessSignature` method shall call its callback with no argument when it succeeds.]*/
+        this._restApiClient.updateSharedAccessSignature(sharedAccessSignature);
+        /*Codes_SRS_NODE_DEVICE_METHOD_CLIENT_16_004: [The `updateSharedAccessSignature` method shall call `updateSharedAccessSignature` on the `RestApiClient` instance used by the `MethodClient` object.]*/
+        callback();
+      }
+    });
+  }
+
+  private _init(callback: (err?: Error) => void): void {
+    this._authProvider.getDeviceCredentials((err, creds) => {
+      if (err) {
+        callback(err);
+      } else {
+        this._httpHeaders['x-ms-edge-moduleId'] = creds.deviceId + '/' + creds.moduleId;
+
+        if (this._restApiClient) {
+          this._restApiClient.updateSharedAccessSignature(creds.sharedAccessSignature);
+          callback();
+        } else {
+          getUserAgentString((userAgentString) => {
+            const transportConfig: RestApiClient.TransportConfig = {
+              host: creds.gatewayHostName,
+              sharedAccessSignature: creds.sharedAccessSignature
+            };
+
+            this._restApiClient = new RestApiClient(transportConfig, userAgentString);
+            this._restApiClient.updateSharedAccessSignature(creds.sharedAccessSignature);
```

Comment:
Is this update call necessary if the token was passed in through the `transportConfig` param?

Reply:
that's a good point...

---
In reply to: [197006491](https://github.com/Azure/azure-iot-sdk-node/pull/322#discussion_r197006491) [](ancestors = 197006491)